File: android/app/src/main/java/com/sapiens/gagaodok/sync/SyncWorkerClient.kt
package com.sapiens.gagaodok.sync

import java.util.Base64
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

data class SyncHttpResponse(val statusCode: Int, val body: ByteArray)
fun interface SyncHttpTransport { fun send(request: Request): SyncHttpResponse }
class OkHttpSyncTransport(private val client: OkHttpClient = OkHttpClient()) : SyncHttpTransport {
    override fun send(request: Request): SyncHttpResponse = client.newCall(request).execute().use {
        SyncHttpResponse(it.code, it.body?.bytes() ?: ByteArray(0))
    }
}
class SyncWorkerClientException(val statusCode: Int? = null) : Exception()

class SyncWorkerClient(
    baseUrl: String,
    /**
     * Read afresh for every request rather than captured once.
     *
     * The settings screen builds its client before enrollment has stored
     * anything, so a token captured at construction would stay empty for the
     * rest of the app session and every read would be refused until a restart.
     */
    private val token: () -> ByteArray?,
    private val transport: SyncHttpTransport = OkHttpSyncTransport(),
) {
    private val baseUrl = baseUrl.removeSuffix("/")
    init {
        require(this.baseUrl.startsWith("https://") && !this.baseUrl.contains('?') && !this.baseUrl.contains('#'))
    }

    /** A client for a token that already exists and will not change. */
    constructor(
        baseUrl: String,
        deviceToken: ByteArray,
        transport: SyncHttpTransport = OkHttpSyncTransport(),
    ) : this(baseUrl, { deviceToken }, transport) {
        require(deviceToken.size == 32)
    }

    /**
     * Built per request from whatever is stored right now.
     *
     * A missing or wrong-sized token is refused here, before a request exists:
     * an unauthenticated request is never sent. The exception carries no
     * status and no value.
     */
    private fun authorization(): String {
        val deviceToken = token()
        if (deviceToken == null || deviceToken.size != 32) throw SyncWorkerClientException()
        return "Device gdt1_" + Base64.getUrlEncoder().withoutPadding().encodeToString(deviceToken)
    }

    @Synchronized fun drainOne(outbox: SyncOutbox): SyncHttpResponse? {
        val entry = outbox.pending().firstOrNull() ?: return null
        val request = request("/v1/sync/operations").post(entry.rawBody.toRequestBody(JSON)).build()
        val response = transport.send(request)
        if (response.statusCode !in 200..299) throw SyncWorkerClientException(response.statusCode)
        outbox.acknowledge(entry.operationId)
        return response
    }
    fun changes(after: Long, limit: Int = 300): SyncHttpResponse = get("/v1/sync/changes?after_seq=$after&limit=$limit")
    fun devices(): SyncHttpResponse = get("/v1/account/devices")
    fun bootstrap(cursor: String? = null, limit: Int = 300): SyncHttpResponse {
        val suffix = cursor?.let { "&cursor=" + java.net.URLEncoder.encode(it, Charsets.UTF_8.name()) } ?: ""
        return get("/v1/sync/bootstrap?limit=$limit$suffix")
    }
    fun putAttachmentContent(attachmentId: String, body: ByteArray): SyncHttpResponse = checked(
        transport.send(
            request("/v1/attachments/${canonicalAttachmentId(attachmentId)}/content")
                .put(body.toRequestBody(OCTET_STREAM)).build(),
        ),
    )

    /**
     * Body 없는 POST다.
     *
     * 예전에 body를 실었더니 원격 complete가 거부해 첨부가 ready가 되지 못했다.
     * local suite가 전부 통과하는 상태였는데도 그랬다.
     */
    fun completeAttachment(attachmentId: String): SyncHttpResponse = checked(
        transport.send(
            request("/v1/attachments/${canonicalAttachmentId(attachmentId)}/complete")
                .post(ByteArray(0).toRequestBody(null)).build(),
        ),
    )

    fun getAttachmentContent(attachmentId: String): SyncHttpResponse = checked(
        transport.send(
            request("/v1/attachments/${canonicalAttachmentId(attachmentId)}/content").get().build(),
        ),
    )

    /** Worker의 경로 매칭이 정규 대문자 UUID만 받는다. */
    private fun canonicalAttachmentId(value: String): String {
        require(value == java.util.UUID.fromString(value).toString().uppercase()) {
            "attachment id must be canonical"
        }
        return value
    }

    private fun checked(response: SyncHttpResponse): SyncHttpResponse {
        if (response.statusCode !in 200..299) throw SyncWorkerClientException(response.statusCode)
        return response
    }

    private fun get(path: String): SyncHttpResponse { val response=transport.send(request(path).get().build());if(response.statusCode !in 200..299)throw SyncWorkerClientException(response.statusCode);return response }
    private fun request(path: String)=Request.Builder().url(baseUrl+path).header("Authorization",authorization()).header("Cache-Control","no-cache")
    companion object {
        private val JSON = "application/json".toMediaType()
        private val OCTET_STREAM = "application/octet-stream".toMediaType()
    }
}

File: Sources/KakaoSapiens/Services/SyncWorkerClient.swift
import Foundation

public struct SyncHTTPResponse {
    public let statusCode: Int
    public let body: Data
}

public protocol SyncHTTPTransport {
    func send(_ request: URLRequest) async throws -> SyncHTTPResponse
}

public struct URLSessionSyncTransport: SyncHTTPTransport {
    private let session: URLSession
    public init(session: URLSession = .shared) { self.session = session }
    public func send(_ request: URLRequest) async throws -> SyncHTTPResponse {
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw SyncWorkerClientError.invalidResponse }
        return SyncHTTPResponse(statusCode: http.statusCode, body: data)
    }
}

public enum SyncWorkerClientError: Error {
    case invalidBaseURL
    case invalidToken
    case invalidResponse
    case httpStatus(Int)
}

public final class SyncWorkerClient {
    private let baseURL: URL
    /// Read afresh for every request rather than captured once.
    ///
    /// The settings screen builds its client before enrollment has stored
    /// anything, so a token captured at construction stays empty for the rest
    /// of the app session and every read is refused until a restart.
    private let token: () -> Data?
    private let transport: SyncHTTPTransport

    public init(baseURL: URL, deviceToken: Data, transport: SyncHTTPTransport = URLSessionSyncTransport()) throws {
        guard baseURL.scheme == "https", baseURL.host != nil, baseURL.query == nil, baseURL.fragment == nil else {
            throw SyncWorkerClientError.invalidBaseURL
        }
        guard deviceToken.count == 32 else { throw SyncWorkerClientError.invalidToken }
        self.baseURL = baseURL
        self.token = { deviceToken }
        self.transport = transport
    }

    /// A client for a token that does not exist yet.
    ///
    /// The provider is consulted per request, so a token stored after this
    /// client was built is used without rebuilding anything. While the
    /// provider has nothing to give the request is refused locally: an
    /// unauthenticated request is never sent.
    public init(
        baseURL: URL,
        token: @escaping () -> Data?,
        transport: SyncHTTPTransport = URLSessionSyncTransport()
    ) throws {
        guard baseURL.scheme == "https", baseURL.host != nil, baseURL.query == nil, baseURL.fragment == nil else {
            throw SyncWorkerClientError.invalidBaseURL
        }
        self.baseURL = baseURL
        self.token = token
        self.transport = transport
    }

    /// Sends exactly one oldest operation. The journal is acknowledged only
    /// after a 2xx response; every error leaves the original bytes untouched.
    public func drainOne(from outbox: SyncOutbox) async throws -> SyncHTTPResponse? {
        guard let entry = try outbox.pending().first else { return nil }
        var request = try authorizedRequest(path: "/v1/sync/operations", method: "POST")
        request.httpBody = entry.rawBody
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let response = try await transport.send(request)
        guard (200..<300).contains(response.statusCode) else {
            throw SyncWorkerClientError.httpStatus(response.statusCode)
        }
        _ = try outbox.acknowledge(operationID: entry.operationID)
        return response
    }

    public func changes(after sequence: UInt64, limit: Int = 300) async throws -> SyncHTTPResponse {
        try await get(path: "/v1/sync/changes?after_seq=\(sequence)&limit=\(limit)")
    }

    public func devices() async throws -> SyncHTTPResponse {
        try await get(path: "/v1/account/devices")
    }

    /// Issue a new recovery phrase for the account this device is already
    /// linked to. The body carries no master key and no entropy.
    public func rotateRecovery(body: Data) async throws -> SyncHTTPResponse {
        var request = try authorizedRequest(path: "/v1/recovery/rotate", method: "POST")
        request.httpBody = body
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let response = try await transport.send(request)
        guard (200..<300).contains(response.statusCode) else {
            throw SyncWorkerClientError.httpStatus(response.statusCode)
        }
        return response
    }

    public func bootstrap(cursor: String? = nil, limit: Int = 300) async throws -> SyncHTTPResponse {
        var path = "/v1/sync/bootstrap?limit=\(limit)"
        if let cursor {
            path += "&cursor=\(cursor.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? "")"
        }
        return try await get(path: path)
    }

    public func putAttachmentContent(attachmentID: String, body: Data) async throws -> SyncHTTPResponse {
        var request = try authorizedRequest(
            path: "/v1/attachments/\(try Self.canonicalAttachmentID(attachmentID))/content", method: "PUT")
        request.httpBody = body
        request.setValue("application/octet-stream", forHTTPHeaderField: "Content-Type")
        return try await transport.send(request)
    }

    /// Body 없는 POST다.
    ///
    /// 예전에 body를 실었더니 원격 `complete`가 거부해 첨부가 `ready`가 되지
    /// 못했다. local suite 877개가 모두 통과하는 상태였는데도 그랬다.
    public func completeAttachment(attachmentID: String) async throws -> SyncHTTPResponse {
        let request = try authorizedRequest(
            path: "/v1/attachments/\(try Self.canonicalAttachmentID(attachmentID))/complete", method: "POST")
        return try await transport.send(request)
    }

    public func getAttachmentContent(attachmentID: String) async throws -> SyncHTTPResponse {
        let request = try authorizedRequest(
            path: "/v1/attachments/\(try Self.canonicalAttachmentID(attachmentID))/content", method: "GET")
        return try await transport.send(request)
    }

    /// Worker의 경로 매칭이 정규 대문자 UUID만 받는다.
    private static func canonicalAttachmentID(_ value: String) throws -> String {
        guard let uuid = UUID(uuidString: value), uuid.uuidString == value else {
            throw SyncWorkerClientError.invalidResponse
        }
        return value
    }

    private func get(path: String) async throws -> SyncHTTPResponse {
        let response = try await transport.send(try authorizedRequest(path: path, method: "GET"))
        guard (200..<300).contains(response.statusCode) else { throw SyncWorkerClientError.httpStatus(response.statusCode) }
        return response
    }

    private func authorizedRequest(path: String, method: String) throws -> URLRequest {
        guard let deviceToken = token(), deviceToken.count == 32 else {
            throw SyncWorkerClientError.invalidToken
        }
        var request = URLRequest(url: URL(string: path, relativeTo: baseURL)!)
        request.httpMethod = method
        request.cachePolicy = .reloadIgnoringLocalCacheData
        request.setValue("Device gdt1_\(deviceToken.base64URLEncoded)", forHTTPHeaderField: "Authorization")
        return request
    }
}

private extension Data {
    var base64URLEncoded: String {
        base64EncodedString().replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_").replacingOccurrences(of: "=", with: "")
    }
}
